centreon/src/Core/HostCategory/Infrastructure/Repository/DbWriteHostCategoryActionLogRepository.php

```diff
@@ -0,0 +1,284 @@
+                }
+
+                return;
+            }
+
+            // If more than one property has changed, we log a change action
+            $actionLog = new ActionLog(
+                objectType: ActionLog::OBJECT_TYPE_HOSTCATEGORIES,
+                objectId: $hostCategory->getId(),
+                objectName: $hostCategory->getName(),
+                actionType: ActionLog::ACTION_TYPE_CHANGE,
+                contactId: $this->user->getId(),
+            );
+            $actionLogId = $this->writeActionLogRepository->addAction($actionLog);
+            $actionLog->setId($actionLogId);
+            $this->writeActionLogRepository->addActionDetails($actionLog, $diff);
+        } catch (\Throwable $ex) {
+            $this->error($ex->getMessage(), ['trace' => $ex->getTraceAsString()]);
+
+            throw $ex;
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function linkToHost(int $hostId, array $categoryIds): void
+    {
+        $this->writeHostCategoryRepository->linkToHost($hostId, $categoryIds);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function unlinkFromHost(int $hostId, array $categoryIds): void
+    {
+        $this->writeHostCategoryRepository->unlinkFromHost($hostId, $categoryIds);
+    }
+
+    /**
+     * Compare the initial and updated HostCategory and return the differences.
+     *
+     * @param HostCategory $initialHostCategory
+     * @param HostCategory $updatedHostCategory
+     *
+     * @return array{
+     *    hc_name?: string,
+     *    hc_alias?: string,
+     *    hc_activate?: string,
+     *    hc_comment?: string
+     * }
+     */
+    private function getHostCategoryDiff(
+        HostCategory $initialHostCategory,
+        HostCategory $updatedHostCategory
+    ): array {
+        $diff = [];
+        $reflection = new \ReflectionClass($initialHostCategory);
+        $properties = $reflection->getProperties();
+
+        foreach ($properties as $property) {
+            $property->setAccessible(true);
+            $value1 = $property->getValue($initialHostCategory);
+            $value2 = $property->getValue($updatedHostCategory);
+
+            if ($value1 !== $value2) {
+                if ($property->getName() === 'isActivated') {
+                    $diff[self::HOST_CATEGORY_PROPERTIES_MAP[$property->getName()]] = $value2 ? '1' : '0';
+                    continue;
+                }
+                if ($property->getName() === 'comment') {
+                    $diff[self::HOST_CATEGORY_PROPERTIES_MAP[$property->getName()]] = $value2 ?? '';
+                    continue;
+                }
+                if ((string) $property->getType() === 'string') {
+                    $diff[self::HOST_CATEGORY_PROPERTIES_MAP[$property->getName()]] = is_string($value2)
+                        ? $value2
+                        : throw new RepositoryException('Property value is not a string');
```

Comment:
Please note that this is not a repository error, but a business rule error for action logs. The decorator should not have the same exception as the repository. It should be an ActionLogException, for example.

Reply:
Seen together, I resolve.